rasp/tor_engine.py

```diff
@@ -1,12 +1,167 @@
+            can also be set with the ``RASP_TOR_CONTROL_PASSWORD``
+            environment variable.
+        control_password (str, optional): Password to protect control
+            port usage,
+            can also be set with the ``RASP_TOR_CONTROL_PORT``
+            environment variable.
+        callback (callable, optional): A function that returns a boolean value
+            to determine when to send a signal to Tor"""
+
+    def __init__(self,
+                 address=None,
+                 port=None,
+                 password=None,
+                 signal_limiter=None):
+        self.address = (
+            address
+            or os.environ.get('RASP_TOR_ADDRESS')
+            or '127.0.0.1'
+        )
+        self.control_port = (
+            port
+            or os.environ.get('RASP_TOR_CONTROL_PORT')
+            or 9051
+        )
+        self.control_password = (
+            password
+            or os.environ.get('RASP_TOR_CONTROL_PASSWORD')
+            or getpass.getpass("Tor control password: ")
+        )
+        self.limiter = signal_limiter
+
+    def __copy__(self):
+        return TorController(
+            address=self.address,
+            port=self.port,
+            password=self.control_password
+        )
+
+    def _enforce_connection(method):
+        """Method decorator to enforce that the connection was established"""
+
+        def enforce(self, *args, **kwargs):
+            if not hasattr(self, 'connection'):
+                raise ControllerError(
+                    'Signal controller has not been opened'
+                )
+            return method(self, *args, **kwargs)
+
+        return enforce
+
+    def open(self):
+        """Establishes a connection to the Tor server"""
+        info = (self.address, self.port)
+        self.connection = connect(
+            control_port=info,
+            password=self.password
+        )
+
+    @_enforce_connection
+    def close(self):
+        """Closes the TorController connection"""
+        self.connection.close()
+
+    @_enforce_connection
+    def ready_to_signal(self):
+        """Checks to see if the Tor server is available to be signalled.
+
+        The availability only applies to the instance's connection, and
+        doesn't take into account any other instances.
+
+        Returns:
+            True if signal can be sent, False if not
+        """
+        return self.connection.is_newnym_available()
+
+    @_enforce_connection
+    def signal(self):
+        """Refreshes the mapping of nodes we connect through.
+
+        When a new path from the user to the destination is required, this
+        method refreshes the node path we use to connect. This gives us a new IP
+        address with which the destination sees.
+        """
+        if self.ready_to_signal():
+            self.connection.signal(Signal.NEWNYM)
+
+    def limited_signal(self):
+        if self.ready_to_signal() and self.limiter():
```

Comment:
should probably wrap this in a try clause, or have a default function for `self.limiter` that always returns `True`.